Refactor Nomad canary releaser into separate package

paladin-devops opened the pull request:
The Nomad `jobspec-canary` releaser plugin is being refactored into a separate plugin from the `jobspec` plugin. Presently, the canary releaser does no harm to any Nomad job deployments which do not have canaries, as the releaser will just exit early. 

However, the default behavior of Waypoint is to "prune" old deployments after a new release has occurred. For "mutable deployment" plugins, such as the `jobspec` plugin, the ID of the deployment is equal to the ID of the Nomad job. This means that even though there is a new version (or "generation") of the Nomad job, deployed by Waypoint, Waypoint by default would delete it right after. [This is called out](https://www.waypointproject.io/plugins/nomad#nomad-jobspec-canary-releasemanager) in the `jobspec-canary` documentation already, and users are recommended to run their `release` operations with the plugin with the `prune=false` flag.

This change is being introduced so that users of the `jobspec` plugin, who have not decided to opt-in to the `jobspec-canary` releaser, are not subject to that behavior. Without it, the `jobspec-canary` plugin is invoked by default, and thus automatic pruning (if not running with the aforementioned prune flag).

Closes #3276 and closes #3316.

paladin-devops committed:
Refactor Nomad canary releaser into separate package.

paladin-devops committed:
Remove canary documentation from Nomad jobspec platform documentation.

paladin-devops committed:
Changelog.

briancain commented on website/content/partials/components/platform-nomad-jobspec-canary.mdx:
```diff
@@ -1,142 +1,11 @@
 ## nomad-jobspec-canary (platform)
 
-Deploy to a Nomad cluster from a pre-existing Nomad job specification file.
```
Do we need to add a new partial mdx file to account for the plugin moving? I think the docs are still around, just not being included because it's a new partial?

briancain replied:
Sweet!

paladin-devops committed:
Remove releaser protobuf from jobspec platform plugin.

paladin-devops committed:
Remove release manager docs for nomad-jobspec platform plugin.

briancain reviewed (approved):
🚀

xiaolin-ninja reviewed (approved):
Looks good! The integration test failure is the same problem from yesterday

catsby commented on builtin/nomad/jobspec/platform.go:
```diff
@@ -74,6 +74,10 @@ func (p *Platform) resourceManager(log hclog.Logger, dcr *component.DeclaredReso
 
+type nomadClient struct {
+	NomadClient *api.Client
+}
```
After doing some digging, I see this comes from the former `releaser.go` file that was in here and both the platform and releaser shared this type. 

That said, the `builtin/nomad` tree now defines `getNomadClient` 4 times: 

```
[waypoint][main](4)$ ack getNomadClient builtin/nomad | grep func

builtin/nomad/task.go:286:func (p *TaskLauncher) getNomadClient() (*api.Client, error) {
builtin/nomad/platform.go:96:func (p *Platform) getNomadClient() (*api.Client, error) {
builtin/nomad/jobspec/releaser.go:73:func (r *Releaser) getNomadClient() (*nomadClient, error) {
builtin/nomad/jobspec/platform.go:79:func (p *Platform) getNomadClient() (*nomadClient, error) {
```

That itself might not be a problem but 2 of them return a `nomadClient` type that wraps the client, and the other 2 do not, and I don't understand why. Any thoughts on consolidating?

catsby replied:
Updated my original comment above to strike the "now" part; it seems before this change the `builtin/nomad` tree already defined the method 4 times :P

paladin-devops replied:
@catsby after taking a look at this, I'm in favor of consolidating the two implementations of this in the `builtin/nomad` package, in `platform.go` and `task.go`. 

I'd like to leave the implementations in `builtin/nomad/jobspec` and `builtin/nomad/canary` separate though, just for the purpose of keeping the plugin packages more independent of each other (given that at some point we may move the built in plugins to their own repos altogether). Though it's not exactly the same situation, I intentionally did not update the `runnerinstall` package to use any utility methods under `builtin`, and likewise did not update anything in `builtin` to depend on the `installutil` package's helper methods, just to keep the plugin code more independent.

Regarding why the client is wrapped, when I initially implemented the resource manager framework for the jobspec & canary plugins earlier this year, I noticed that the k8s resource manager implementation wrapped the k8s client set in a struct. Albeit in that situation, the clientset, namespace and client config are provided in that struct (copied below for reference):

```go
// getClientset is a value provider for our resource manager and provides
// the connection information used by resources to interact with Kubernetes.
func (p *Platform) getClientset() (*clientsetInfo, error) {
	// Get our client
	clientSet, ns, config, err := Clientset(p.config.KubeconfigPath, p.config.Context)
	if err != nil {
		return nil, err
	}

	return &clientsetInfo{
		Clientset: clientSet,
		Namespace: ns,
		Config:    config,
	}, nil
}
```

Given that, since for Nomad we only need `*api.Client`, I think that `getNomadClient` in the jobspec and canary packages can be updated to return just that, instead of a struct wrapping it.

catsby replied:
Sounds rational, thanks 👍

catsby reviewed (changes requested):
I'd like to understand and document any backwards compatibility concerns here before we proceed.

catsby reviewed (changes requested):
Please do not merge this until a patch for https://github.com/hashicorp/waypoint/issues/3601 is merged and released

paladin-devops committed:
Update .changelog/3359.txt

Changelog now reflects that the Nomad plugin is being updated instead of specifically the Nomad jobspec plugin, and clarifies what the change is from a user-facing perspective.

Co-authored-by: Clint <catsby@users.noreply.github.com>

paladin-devops commented on builtin/nomad/jobspec/plugin.proto:
```diff
@@ -12,11 +12,6 @@ message Deployment {
-message Release {
-  string Url = 1;
-  opaqueany.Any resource_state = 2;
-}
```
@catsby regarding backwards-compatibility, a waypoint.hcl file using the `nomad-jobspec-canary` releaser will work as-is with this change. However, I have not tested the path where a deployment was created with 0.8.x, and then attempted to be destroyed with this branch. I can try that out, but I'm okay with un-backporting this.

evanphx replied:
Deleting the definition won't break anything. I presume (but don't see?) that nothing is using this message as a type in another message anymore.

With protobuf, changing the type of existing fields is the number 1 backward compat issue. Deleting types just means you can't decode them in the future (but since you deleted the type, it's assumed you don't need to decode them either).

paladin-devops committed:
Consolidate getNomadClient function for Nomad platform and task launcher.

paladin-devops committed:
Update getNomadClient for jobspec & canary plugins to return *api.Client instead of a struct wrapping that.

briancain commented:
I think this is good to merge now @paladin-devops ?

paladin-devops commented:
@briancain this should be good to go! I just rebased again, thinking that my last commit failed the integration test & needed the updates from main, but it was just the failed Percy check 😄 after the checks run it should be all set.
